File: leetcode/785_Is_Graph_Bipartite.py
from typing import List
from unittest import TestCase

class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        colours = [None] * n
        for i in range(n):
            if colours[i] is None:
                stack, colours[i] = [i], 0
                while stack:
                    v = stack.pop()
                    for u in graph[v]:
                        if colours[u] is None:
                            colours[u] = not colours[v]
                            stack.append(u)
                        elif colours[v] == colours[u]:
                            return False
        return True

class IsBipartiteTest(TestCase):
    def test_1(self) -> None:
        self.assertFalse(Solution().isBipartite([[1,2,3],[0,2],[0,1,3],[0,2]]))

    def test_2(self) -> None:
        self.assertTrue(Solution().isBipartite([[1,3],[0,2],[1,3],[0,2]]))

File: leetcode/785_Is_Graph_Bipartite.rs
struct Solution {}

impl Solution {
    pub fn is_bipartite(graph: Vec<Vec<i32>>) -> bool {
        let n = graph.len();
        let mut colours = vec![None; n];
        let mut stack = Vec::with_capacity(n);
        for i in 0..n {
            if colours[i].is_none() {
                stack.push(i);
                colours[i] = Some(true);
                while let Some(v) = stack.pop() {
                    for u in graph[v].iter().map(|x| *x as usize) {
                        if colours[u].is_none() {
                            stack.push(u);
                            colours[u] = colours[v].map(|x| !x);
                        } else if colours[u] == colours[v] { 
                            return false;
                        }
                    }
                }
            }
        }
        true
    }
}

#[test]
fn test_1() {
    assert!(! Solution::is_bipartite(vec![vec![1,2,3], vec![0,2], vec![0,1,3], vec![0,2]]));
}

#[test]
fn test_2() {
    assert!(Solution::is_bipartite(vec![vec![1,3], vec![0,2], vec![1,3], vec![0,2]]));
}

